Add benchmarks from unyt paper (Goldbaum 2018)

pllim opened the pull request:
Fix #53. xref astropy/astropy#7546 and astropy/astropy#7549

pllim commented on benchmarks/units.py:
```diff
@@ -68,6 +70,22 @@ def time_quantity_init_array():
+    https://github.com/astropy/astropy/issues/7546 reported high overhead
+    for small list.
+    """
+    b1 * u.m / u.s
```
Note: This is not exactly as done at https://github.com/yt-project/unyt/blob/master/benchmarks/bench.py but it is more consistent with existing benchmarks for unit initialization.

pllim commented on benchmarks/units.py:
```diff
@@ -86,3 +104,64 @@ def time_quantity_times_quantity():
+        self.out = data.copy()
+
+        # A different but dimensionally compatible unit
+        self.data2 = 0.001 * data * u.kg
```
Note: In the paper, these benchmarks were also done for `data2` having same unit as `data1`, but I did not include it here. My reasoning is that `data2` having a different unit is a worse case scenario, hence it is unnecessary to include the other one. Please correct me if I am wrong.

ngoldbaum replied:
I included both since it triggers different code paths (and in particular a possible early return if the units are identical).

pllim replied:
@mhvk , what are your thoughts here and for the comment below this one? Do we need to benchmark all possible routes?

mhvk replied:
I think it is probably good to include the same-unit case - we do indeed try to make that faster, so if it suddenly becomes a lot slower, it would be good to have warning.

pllim commented on benchmarks/units.py:
```diff
@@ -86,3 +104,64 @@ def time_quantity_times_quantity():
+        np.sqrt(self.data, out=self.out)
+
+    def time_quantity_equal(self):
+        self.data == self.data2
```
Note: In the paper, `np.equal` and `operator.eq` are compared separately. I did not do that here. Is that necessary here or is `==` alone sufficient? Same note for all the following operator benchmarks. Please let me know.

ngoldbaum replied:
operator.eq and `==` are equivalent, so there's no reason to test both of those. Going through `np.equal` with short-circuit `__eq__` if that's defined, so that might be different.

pllim commented on benchmarks/units.py:
```diff
@@ -86,3 +104,64 @@ def time_quantity_times_quantity():
+        self.data + self.data2
+
+
+class TimeQuantityOpLargeArray(TimeQuantityOpSmallArray):
```
@astrofrog , does subclassing like this work for `asv`?

mhvk replied:
If the subclassing does work, that could also be done for the same-unit case.

ngoldbaum commented:
This is probably a separate issue, but don't astropy's benchmarks use airspeed velocity? The last time I checked asv doesn't have a way to control CPU throttling or other effects that might introduce benchmark variance. You may want to consider integrating `perf` into your airspeed velocity setup to reduce issues like that.

pllim commented:
> You may want to consider integrating `perf` into your airspeed velocity setup

Perhaps @astrofrog can comment on this. Locally, I run it on a single core on my machine but I don't know my OS does throttling or not. I think we see small scale variations but overall trend is reliable. Ref: http://www.astropy.org/astropy-benchmarks/

astrofrog commented:
@ngoldbaum:

> This is probably a separate issue, but don't astropy's benchmarks use airspeed velocity? The last time I checked asv doesn't have a way to control CPU throttling or other effects that might introduce benchmark variance

When running asv, we only run it on physical machines (not using CI) and prevent core-swapping using ``taskset -c 0 asv`` so we get reasonably stable results:

http://www.astropy.org/astropy-benchmarks/#coordinates.time_latitude

pllim commented:
@mhvk and @ngoldbaum , I think I addressed all your comments. 

@astrofrog , I ran the benchmarks and results should be in https://github.com/pllim/astropy-benchmarks/tree/lim-test but I am having trouble getting them to show up in https://pllim.github.io/astropy-benchmarks/ . I think it is the same problem as #50 but I am not sure where is the script that you updated for the fix.

pllim commented:
Update: Turns out I needed to also copy `results/benchmarks.json` over to "results" branch, not just `results/machine_name/` folder. New benchmarks (for limited number of commits) appear at https://pllim.github.io/astropy-benchmarks , so subclassing seems to work!
